### PR TITLE
force slave for columns/table_exist so we do not go to the master eve…

### PR DESCRIPTION
**Pull request by grosser**
…n when in an transaction

more extreme version of https://github.com/zendesk/active_record_shards/pull/112
since I keep running into this during debug sessions and it seems like an easy win ...

### CODE REVIEW

**Comment by grosser** [comment]
updated and tested!

**Comment by libo** [comment]
🎉 Shall we merge it and brig it out?

**Comment by libo** [comment]
@bquorning can you please help me review this tomorrow? Thanks!

File: lib/active_record_shards/connection_switcher.rb [hunk 1]
```diff
@@ -95,18 +95,18 @@ def on_master(&block)
     alias_method :with_slave_if, :on_slave_if
     alias_method :with_slave_unless, :on_slave_unless
 
-    def on_cx_switch_block(which, options = {}, &block)
+    def on_cx_switch_block(which, force: false, construct_ro_scope: nil, &block)
```
Comment (libo) [inline lib/active_record_shards/connection_switcher.rb]: You are changing the signature of a public method. This is kind of an API change.

File: lib/active_record_shards/default_slave_patches.rb [hunk 2]
```diff
@@ -25,15 +25,16 @@ def #{method}_with_default_slave#{punctuation}(*args, &block)
       RUBY
     end
 
-    def columns_with_default_slave(*args, &block)
-      read_columns_from =
-        if on_slave_by_default? && !Thread.current[:_active_record_shards_slave_off]
```
Comment (libo) [inline lib/active_record_shards/default_slave_patches.rb]: Don't we need to check the current Thread anymore?

File: lib/active_record_shards/connection_switcher.rb (hunk 1 above)
Comment (bquorning) [reply lib/active_record_shards/connection_switcher.rb]: Options hash is equivalent to keyword arguments. But kwargs are more explicit. Not an API change imho.

Comment (libo) [reply lib/active_record_shards/connection_switcher.rb]: true that

**Comment by grosser** [comment]
only failure could be when someone passes a hash with more options ... but
that's exactly what I want to catch :)

File: lib/active_record_shards/default_slave_patches.rb (hunk 2 above)
Comment (grosser) [reply lib/active_record_shards/default_slave_patches.rb]: nope ... forcing it even if inside a thread

**Comment by bquorning** [comment]
The original implementation of `#columns_with_default_slave` was added in 7a10741628c1587fd85ddb5dafedba66253a82d3 with a commit message saying something about “fixing the stack overflow in `.columns`”. Do you know which overflows that might refer to?

**Comment by grosser** [comment]
the `construct_ro_scope: false` is still there ... so I think it should stay stackoverflow-free

**Pull request merged**